File: 03-adam.py
import torch

N, D_in, H, D_out = 64, 1000, 100, 10

x = torch.randn(N, D_in, requires_grad=False)
y = torch.randn(N, D_out, requires_grad=False)

w1 = torch.randn(D_in, H, requires_grad=True)
w2 = torch.randn(H, D_out, requires_grad=True)

learning_rate = 1e-3
adam = torch.optim.Adam([w1, w2], lr=learning_rate)

for t in range(500):
    y_pred = x.mm(w1).clamp(min=0).mm(w2)
    loss = (y_pred - y).pow(2).sum()

    if t % 100 == 99:
        print(t, loss.item())

    adam.zero_grad()
    loss.backward()
    adam.step()

File: 03-adam.go
package main

import (
	"fmt"

	at "github.com/gotorch/gotorch/aten"
	"github.com/gotorch/gotorch/torch"
	"github.com/gotorch/gotorch/torch/optim"
)

func main() {
	N, D_in, H, D_out := 64, 1000, 100, 10
	learning_rate := 1e-3

	x := torch.RandN([]int{N, Din},
		at.TensorOptions().RequiresGrad(false))
	y := torch.RandN([]int{N, Dout},
		at.TensorOptions().RequiresGrad(false))

	params := []at.Tensor{
		torch.RandN([]int{Din, H},
			at.TensorOptions().RequiresGrad(true)),
		torch.RandN([]int{H, Dout},
			at.TensorOptions().RequiresGrad(true)),
	}

	adam := optim.NewAdam(params, optim.AdamOptions(learning_rate))

	w1 := adam.parameters()[0]
	w2 := adam.parameters()[1]

	for i := 0; i < 500; i++ {
		y_pred := at.Sum(at.Clamp(at.MM(x, w1), 0), w2)
		loss := at.Sum(at.Pow(at.Sub(y_pred, y), 2))

		if i%100 == 0 {
			fmt.Println("loss = ", loss)
		}

		adam.ZeroGrad()
		loss.Backward()
		adam.Step()
	}
}
